File: framework/dotnet/BetterPortal/Elevation.cs
using Node = System.Collections.Generic.Dictionary<string, object?>;
namespace BetterPortal;

public sealed class ElevationRequiredException : TokenException
{
    public IReadOnlyDictionary<string, string> Headers { get; }
    public ElevationRequiredException(Node requirement, Node claims) : base("insufficient_user_authentication")
    {
        var challenge = new Node(requirement) { ["version"] = 1, ["tenantId"] = claims["tenantId"], ["appId"] = claims["appId"] };
        var acr = Equals(requirement["minimum"], "mfa") ? "mfa" : "confirmed";
        var authenticate = $"Bearer error=\"insufficient_user_authentication\", acr_values=\"urn:betterportal:{acr}\"";
        if (requirement.TryGetValue("maxAgeSeconds", out var age)) authenticate += $", max_age=\"{age}\"";
        Headers = new Dictionary<string, string> { ["WWW-Authenticate"] = authenticate, ["BP-Auth-Challenge"] = Json.Write(challenge), ["Cache-Control"] = "no-store" };
    }
}
public static class Elevation
{
    /// <summary>Apply to verified app-bound claims after permissions, before side effects.</summary>
    public static void Require(Node? claims, Node requirement, long? now = null)
    {
        var policy = (Node)Contracts.Parse("ElevationRequirementSchema", requirement)!;
        if (claims is null) throw new TokenException("Human authentication required");
        var current = now ?? DateTimeOffset.UtcNow.ToUnixTimeSeconds();
        if (claims.GetValueOrDefault("elevation") is Node e && Equals(claims.GetValueOrDefault("tokenType"), "access")) {
            var verifiedAt = Convert.ToInt64(e.GetValueOrDefault("verifiedAt", -1));
            var expiresAt = Convert.ToInt64(e.GetValueOrDefault("expiresAt", 0));
            var expires = Convert.ToInt64(claims["exp"]);
            if (e.GetValueOrDefault("assurance") is "confirmed" or "mfa" && verifiedAt >= 0 && verifiedAt <= current && current < expiresAt && expiresAt <= expires
                && (!Equals(policy["minimum"], "mfa") || Equals(e["assurance"], "mfa"))
                && (!policy.TryGetValue("maxAgeSeconds", out var age) || current - verifiedAt <= Convert.ToInt64(age))) return;
        }
        throw new ElevationRequiredException(policy, claims);
    }
}

File: framework/python/betterportal/elevation.py
"""Step-up checks for already verified, app-bound access claims."""
from __future__ import annotations
import json
import time
from typing import Any, Mapping
from .contracts import parse
from .security import TokenError


class ElevationRequired(TokenError):
    def __init__(self, requirement: Mapping[str, Any], claims: Mapping[str, Any]):
        super().__init__("insufficient_user_authentication")
        self.requirement = dict(requirement)
        challenge = {"version": 1, "tenantId": claims["tenantId"], "appId": claims["appId"], **requirement}
        acr = "mfa" if requirement["minimum"] == "mfa" else "confirmed"
        authenticate = f'Bearer error="insufficient_user_authentication", acr_values="urn:betterportal:{acr}"'
        if "maxAgeSeconds" in requirement: authenticate += f', max_age="{requirement["maxAgeSeconds"]}"'
        self.headers = {"WWW-Authenticate": authenticate, "BP-Auth-Challenge": json.dumps(challenge, separators=(",", ":")), "Cache-Control": "no-store"}


def require_elevation(claims: Mapping[str, Any] | None, requirement: Mapping[str, Any], *, now: int | None = None) -> None:
    """Check after authorization and before effects. Never accepts unverified browser claims."""
    policy = parse("ElevationRequirementSchema", dict(requirement))
    if claims is None: raise TokenError("Human authentication required")
    current = int(time.time()) if now is None else now
    e = claims.get("elevation")
    if (isinstance(e, dict) and claims.get("tokenType") == "access" and claims["exp"] > current
        and e.get("assurance") in ("confirmed", "mfa")
        and 0 <= e.get("verifiedAt", -1) <= current < e.get("expiresAt", 0) <= claims["exp"]
        and (policy["minimum"] != "mfa" or e["assurance"] == "mfa")
        and ("maxAgeSeconds" not in policy or current - e["verifiedAt"] <= policy["maxAgeSeconds"])): return
    raise ElevationRequired(policy, claims)
